File: rag/embeddings.py
"""
Lightweight embedding class using HuggingFace Inference API.
No PyTorch, no sentence-transformers — just HTTP calls.
Uses the same all-MiniLM-L6-v2 model as our Pinecone index.
"""
import os
import requests

HF_MODEL = "sentence-transformers/all-MiniLM-L6-v2"
HF_API_URL = f"https://router.huggingface.co/hf-inference/pipeline/feature-extraction/{HF_MODEL}"


class HFInferenceEmbedding:
    """Lightweight embedding using HuggingFace Inference API (free tier with token)."""

    def __init__(self):
        self.model_name = HF_MODEL

    def _get_headers(self) -> dict:
        headers = {"Content-Type": "application/json"}
        hf_token = os.getenv("HF_TOKEN")
        if hf_token:
            headers["Authorization"] = f"Bearer {hf_token}"
        return headers

    def embed(self, texts: list[str]) -> list[list[float]]:
        """Call HuggingFace Inference API for embeddings."""
        response = requests.post(
            HF_API_URL,
            headers=self._get_headers(),
            json={"inputs": texts, "options": {"wait_for_model": True}},
            timeout=30,
        )
        response.raise_for_status()
        return response.json()

    def embed_query(self, query: str) -> list[float]:
        """Embed a single query string."""
        return self.embed([query])[0]

File: ceo-arena-next/lib/rag/embeddings.ts
const HF_MODEL = "sentence-transformers/all-MiniLM-L6-v2";
const HF_API_URL = `https://router.huggingface.co/hf-inference/pipeline/feature-extraction/${HF_MODEL}`;

export class HFInferenceEmbedding {
  private getHeaders(): Record<string, string> {
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    const hfToken = process.env.HF_TOKEN;
    if (hfToken) {
      headers["Authorization"] = `Bearer ${hfToken}`;
    }
    return headers;
  }

  async embed(texts: string[]): Promise<number[][]> {
    const response = await fetch(HF_API_URL, {
      method: "POST",
      headers: this.getHeaders(),
      body: JSON.stringify({ inputs: texts, options: { wait_for_model: true } }),
      signal: AbortSignal.timeout(30000),
    });

    if (!response.ok) {
      throw new Error(`HF Inference API error: ${response.status} ${response.statusText}`);
    }

    return response.json() as Promise<number[][]>;
  }

  async embedQuery(query: string): Promise<number[]> {
    const result = await this.embed([query]);
    const first = result[0];
    if (!first) throw new Error("Empty embedding result");
    return first;
  }
}
